File: backend/app/api/ltv.py
"""LTV calculator endpoints (LP-77) — mirrors the DTI calculator's endpoints.

``GET`` returns the full, transparent, auto-populated calculation (the three
ratios + the itemized breakdown); ``PUT`` / ``DELETE`` set and clear a per-field
override, recomputing in the response (the real-time recalc). Every route is
tenant-scoped (cross-company → 404). Overrides are audited in the service.
"""

from fastapi import APIRouter, HTTPException, status

from app.api.dependencies import CurrentUser
from app.core.database import DbSession
from app.schemas.ltv import LtvCalculation, LtvOverrideInput
from app.services.loan_files import get_loan_file
from app.services.ltv import (
    UnknownLtvFieldError,
    build_ltv_calculation,
    clear_ltv_override,
    set_ltv_override,
)

router = APIRouter(prefix="/loan-files", tags=["ltv"])

_NOT_FOUND = HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Loan file not found")


@router.get("/{identifier}/ltv", response_model=LtvCalculation)
async def get_ltv(identifier: str, db: DbSession, current_user: CurrentUser) -> LtvCalculation:
    """The auto-populated, itemized LTV calculation for one of the caller's files."""
    loan_file = await get_loan_file(db, company_id=current_user.company_id, identifier=identifier)
    if loan_file is None:
        raise _NOT_FOUND
    return await build_ltv_calculation(db, loan_file=loan_file)


@router.put("/{identifier}/ltv/overrides/{field_key}", response_model=LtvCalculation)
async def put_ltv_override(
    identifier: str,
    field_key: str,
    payload: LtvOverrideInput,
    db: DbSession,
    current_user: CurrentUser,
) -> LtvCalculation:
    """Override one LTV input (audited, persisted) and return the recomputed result."""
    loan_file = await get_loan_file(db, company_id=current_user.company_id, identifier=identifier)
    if loan_file is None:
        raise _NOT_FOUND
    try:
        calculation = await set_ltv_override(
            db,
            loan_file=loan_file,
            field_key=field_key,
            data=payload,
            actor_user_id=current_user.id,
        )
    except UnknownLtvFieldError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Unknown LTV input field"
        ) from exc
    await db.commit()
    return calculation


@router.delete("/{identifier}/ltv/overrides/{field_key}", response_model=LtvCalculation)
async def delete_ltv_override(
    identifier: str, field_key: str, db: DbSession, current_user: CurrentUser
) -> LtvCalculation:
    """Clear an override (revert to the auto value), audited; return the recompute."""
    loan_file = await get_loan_file(db, company_id=current_user.company_id, identifier=identifier)
    if loan_file is None:
        raise _NOT_FOUND
    calculation = await clear_ltv_override(
        db, loan_file=loan_file, field_key=field_key, actor_user_id=current_user.id
    )
    await db.commit()
    return calculation

File: frontend/lib/api/ltv.ts
/**
 * LTV calculator data layer (LP-77) — mirrors the DTI data layer (LP-76).
 *
 * The override endpoints return the recomputed calculation, so a mutation primes
 * the query cache directly (`setQueryData`) — real-time recalc from one
 * round-trip. The activity feed is invalidated so the audited override appears.
 */
import { apiClient } from "@/lib/api/client";
import type { LtvCalculation, LtvOverrideInput } from "@/lib/types/ltv";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { isAxiosError } from "axios";

const API_V1 = "/api/v1";

export const ltvQueryKey = (identifier: string) => ["ltv", identifier] as const;

export async function fetchLtv(identifier: string): Promise<LtvCalculation> {
  const res = await apiClient.get<LtvCalculation>(`${API_V1}/loan-files/${identifier}/ltv`);
  return res.data;
}

/** A 404 (missing or out-of-company) won't change on retry — surface it. */
function noRetryOn404(failureCount: number, error: unknown): boolean {
  return !(isAxiosError(error) && error.response?.status === 404) && failureCount < 1;
}

export function useLtv(identifier: string) {
  return useQuery({
    queryKey: ltvQueryKey(identifier),
    queryFn: () => fetchLtv(identifier),
    enabled: Boolean(identifier),
    retry: noRetryOn404,
  });
}

export async function setLtvOverride(
  identifier: string,
  fieldKey: string,
  input: LtvOverrideInput,
): Promise<LtvCalculation> {
  const res = await apiClient.put<LtvCalculation>(
    `${API_V1}/loan-files/${identifier}/ltv/overrides/${fieldKey}`,
    input,
  );
  return res.data;
}

export async function clearLtvOverride(
  identifier: string,
  fieldKey: string,
): Promise<LtvCalculation> {
  const res = await apiClient.delete<LtvCalculation>(
    `${API_V1}/loan-files/${identifier}/ltv/overrides/${fieldKey}`,
  );
  return res.data;
}

/** Set an override → prime the cache with the recomputed result (real-time recalc). */
export function useSetLtvOverride(identifier: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ fieldKey, input }: { fieldKey: string; input: LtvOverrideInput }) =>
      setLtvOverride(identifier, fieldKey, input),
    onSuccess: (data) => {
      queryClient.setQueryData(ltvQueryKey(identifier), data);
      void queryClient.invalidateQueries({ queryKey: ["loan-file-activity", identifier] });
    },
  });
}

/** Clear an override (revert to the auto value) → prime the cache with the recompute. */
export function useClearLtvOverride(identifier: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (fieldKey: string) => clearLtvOverride(identifier, fieldKey),
    onSuccess: (data) => {
      queryClient.setQueryData(ltvQueryKey(identifier), data);
      void queryClient.invalidateQueries({ queryKey: ["loan-file-activity", identifier] });
    },
  });
}
